File: ref/algorithmic-trading/apps/app/app/src/main/java/com/trading/app/domain/order/OrderStateMachine.kt
package com.trading.app.domain.order

import com.trading.app.domain.model.OrderStatus

typealias OnOrderTransition = (from: OrderStatus, to: OrderStatus) -> Unit

class OrderStateMachine(
    initialStatus: OrderStatus = OrderStatus.CREATED,
    private val onTransition: OnOrderTransition? = null,
) {
    private var _status: OrderStatus = initialStatus

    val status: OrderStatus get() = _status

    val isTerminal: Boolean
        get() = _status in TERMINAL_STATES

    fun canTransition(to: OrderStatus): Boolean {
        val allowed = TRANSITIONS[_status] ?: return false
        return to in allowed
    }

    fun transition(to: OrderStatus) {
        if (!canTransition(to)) {
            throw IllegalStateException("Invalid order state transition: ${_status} → $to")
        }
        val from = _status
        _status = to
        onTransition?.invoke(from, to)
    }

    fun allowedTransitions(): List<OrderStatus> =
        TRANSITIONS[_status] ?: emptyList()

    companion object {
        private val TRANSITIONS: Map<OrderStatus, List<OrderStatus>> = mapOf(
            OrderStatus.CREATED to listOf(OrderStatus.SUBMITTED, OrderStatus.REJECTED, OrderStatus.ERROR),
            OrderStatus.SUBMITTED to listOf(OrderStatus.PENDING, OrderStatus.REJECTED),
            OrderStatus.PENDING to listOf(
                OrderStatus.FILLED, OrderStatus.PARTIAL_FILLED,
                OrderStatus.REJECTED, OrderStatus.CANCEL_REQUESTED,
            ),
            OrderStatus.PARTIAL_FILLED to listOf(
                OrderStatus.FILLED, OrderStatus.CANCEL_REQUESTED,
            ),
            OrderStatus.FILLED to emptyList(),
            OrderStatus.REJECTED to emptyList(),
            OrderStatus.CANCEL_REQUESTED to listOf(OrderStatus.CANCELLED, OrderStatus.FILLED),
            OrderStatus.CANCELLED to emptyList(),
        )

        private val TERMINAL_STATES: Set<OrderStatus> = setOf(
            OrderStatus.FILLED, OrderStatus.REJECTED, OrderStatus.CANCELLED, OrderStatus.ERROR,
        )
    }
}

File: ref/algorithmic-trading/apps/engine/src/state/order-state-machine.ts
import { OrderStatus } from '@trading/shared/types';

/** 주문 상태 전이 규칙 */
const TRANSITIONS: Record<OrderStatus, OrderStatus[]> = {
  [OrderStatus.CREATED]: [OrderStatus.SUBMITTED],
  [OrderStatus.SUBMITTED]: [OrderStatus.PENDING, OrderStatus.REJECTED],
  [OrderStatus.PENDING]: [
    OrderStatus.FILLED,
    OrderStatus.PARTIAL_FILLED,
    OrderStatus.REJECTED,
    OrderStatus.CANCEL_REQUESTED
  ],
  [OrderStatus.PARTIAL_FILLED]: [
    OrderStatus.FILLED,
    OrderStatus.CANCEL_REQUESTED
  ],
  [OrderStatus.FILLED]: [],
  [OrderStatus.REJECTED]: [],
  [OrderStatus.CANCEL_REQUESTED]: [OrderStatus.CANCELLED, OrderStatus.FILLED],
  [OrderStatus.CANCELLED]: []
};

/** 종료 상태 (더 이상 전이 불가) */
const TERMINAL_STATES: Set<OrderStatus> = new Set([
  OrderStatus.FILLED,
  OrderStatus.REJECTED,
  OrderStatus.CANCELLED
]);

/** 상태 전이 콜백 */
export type OnOrderTransition = (from: OrderStatus, to: OrderStatus) => void;

export class OrderStateMachine {
  private _status: OrderStatus;
  private _onTransition?: OnOrderTransition;

  constructor(
    initialStatus: OrderStatus = OrderStatus.CREATED,
    onTransition?: OnOrderTransition
  ) {
    this._status = initialStatus;
    this._onTransition = onTransition;
  }

  get status(): OrderStatus {
    return this._status;
  }

  get isTerminal(): boolean {
    return TERMINAL_STATES.has(this._status);
  }

  /** 전이 가능 여부 확인 */
  canTransition(to: OrderStatus): boolean {
    const allowed = TRANSITIONS[this._status];
    return allowed !== undefined && allowed.includes(to);
  }

  /** 상태 전이 실행. 불가능하면 에러 */
  transition(to: OrderStatus): void {
    if (!this.canTransition(to)) {
      throw new Error(
        `Invalid order state transition: ${this._status} → ${to}`
      );
    }
    const from = this._status;
    this._status = to;
    this._onTransition?.(from, to);
  }

  /** 허용된 다음 상태 목록 */
  allowedTransitions(): OrderStatus[] {
    return TRANSITIONS[this._status] ?? [];
  }
}
